yarn-project/prover-node/src/prover-coordination/combined-prover-coordination.ts

```diff
@@ -0,0 +1,147 @@
+export class CombinedProverCoordination<T extends P2PClientType = P2PClientType.Full> implements ProverCoordination {
+  constructor(
+    public readonly p2p: P2P<T> | undefined,
+    public readonly aztecNodes: TxSource[],
+    private readonly options: CombinedCoordinationOptions = {
+      txGatheringBatchSize: 10,
+      txGatheringMaxParallelRequests: 10,
+      txGatheringTimeoutMs: 10000,
+      txGatheringIntervalMs: 1000,
+    },
+    private readonly log = createLogger('aztec:prover-node:combined-prover-coordination'),
+  ) {}
+
+  public async getTxsByHash(txHashes: TxHash[]): Promise<Tx[]> {
+    const pool = this.p2p ? new P2PCoordinationPool(this.p2p) : new InMemoryCoordinationPool();
+    await this.#gatherTxs(txHashes, pool);
+    const notFound = await pool.getUnavailableTxsFromPool(txHashes);
+    if (notFound.length > 0) {
+      throw new Error(`Could not find txs: ${notFound.map(tx => tx.toString())}`);
+    }
+    const txs = await pool.getTxsByHashFromPool(txHashes);
+    return txs.filter(tx => tx !== undefined) as Tx[];
+  }
+
+  public async gatherTxs(txHashes: TxHash[]): Promise<void> {
+    const pool = this.p2p ? new P2PCoordinationPool(this.p2p) : new InMemoryCoordinationPool();
+    await this.#gatherTxs(txHashes, pool);
+  }
+
+  async #gatherTxs(txHashes: TxHash[], pool: CoordinationPool): Promise<void> {
+    const notFound = await pool.getUnavailableTxsFromPool(txHashes);
+    const txsToFind = new Set(notFound.map(tx => tx.toString()));
+    if (txsToFind.size === 0) {
+      return;
+    }
+    this.log.info(`Gathering ${txsToFind.size} txs from any available nodes`);
+    await this.#gatherTxsFromAllNodes(txsToFind, pool);
+    if (txsToFind.size === 0) {
+      return;
+    }
+    this.log.info(`Gathering ${txsToFind.size} txs from p2p network`);
+    await pool.getTxsByHash([...txsToFind].map(tx => TxHash.fromString(tx)));
+  }
+
+  async #gatherTxsFromAllNodes(txsToFind: Set<string>, pool: CoordinationPool) {
+    if (txsToFind.size === 0 || this.aztecNodes.length === 0) {
+      return;
+    }
+    await Promise.all(this.aztecNodes.map(aztecNode => this.#gatherTxsFromNode(txsToFind, aztecNode, pool)));
+  }
+
+  async #gatherTxsFromNode(txsToFind: Set<string>, aztecNode: TxSource, pool: CoordinationPool) {
+    const totalTxsRequired = txsToFind.size;
+
+    // It's possible that the set is empty as we already found the txs
+    if (totalTxsRequired === 0) {
+      return;
+    }
+    let totalTxsGathered = 0;
+
+    const batches: string[][] = [];
+    const allTxs: string[] = [...txsToFind];
+    while (allTxs.length) {
+      const batch = allTxs.splice(0, this.options.txGatheringBatchSize);
+      batches.push(batch);
+    }
+
+    await asyncPool(this.options.txGatheringMaxParallelRequests, batches, async batch => {
+      try {
+        const txs = (await aztecNode.getTxsByHash(batch.map(b => TxHash.fromString(b)))).filter(
+          tx => tx !== undefined,
+        ) as Tx[];
```

Comment:
You can do something like `txs.filter((tx): tx is Tx => !!tx)` to avoid the cast